pyteal/compiler/compiler_test.py

```diff
@@ -2193,3 +2193,448 @@ def access_b4_store(magic_num: pt.abi.Uint64, *, output: pt.abi.Uint64):
 
     with pytest.raises(pt.TealInternalError):
         pt.compileTeal(program_access_b4_store_broken, pt.Mode.Application, version=6)
+
+
+def test_router_app():
+    router = pt.Router()
+
+    router.add_bare_call(pt.Approve(), pt.OnComplete.ClearState)
```

Comment:
The API becomes cleaner if it is purely decorator driven with only an `__init__()` at the beginning and a `compile()` at the end. One way to acheive this is to allow `__init__()` to take the bare calls as a param with good defaults, e.g.:

```python
# before Router:
@dataclass
class BareCall:
    on_complete: OnComplete
    action: ABIReturnSubroutine | SubroutineFnWrapper | Expr
    creation: bool = False

# there may be more defaults to choose from, I'm just copy-pasta'ing what I see here:
# https://github.com/algorand-devrel/wormhole-demo/blob/main/demos/pyth/contracts/contract.py#L7
DEFAULT_BARE_CALLS: list[BareCall] = [
    BareCall(OnComplete.NoOp, Approve(), True),
    BareCall(OnComplete.DeleteApplication, Return(Txn.sender() == Global.creator_address())),
    BareCall(OnComplete.UpdateApplication, Return(Txn.sender() == Global.creator_address()),
    BareCall(OnComplete.OptIn, Reject()),
    BareCall(OnComplete.CloseOut, Reject()),
    BareCall(OnComplete.ClearState, Reject()),
]
 ...
# inside Router:
    def __init__(
        self,
        name: Optional[str] = None,
        bare_calls: list[BareCall] = DEFAULT_BARECALLS,
        # not sure how this works together with the current params of type `list[ProgramNode]`
        ...
    ) -> None:
        self.name: str = "Contract" if name is None else name
        for bare_call in bare_calls:
            self.add_bare_call(bare_call=bare_call) # assuming a new param `bare_call` with other defaults
        ...
```

Reply:
good point on a cleaner API suggestion! just a thing on example syntax: seems `on_complete` should be either an element or a list of on_completes.

If I keep walking down this line, the question is that, do we actually need to expose `add_bare_call` as a method? Since you are able to add all bare calls on `__init__`...

Reply:
> It is easier to add functionality to a public API rather than remove.

good point, designing a reasonable API is not easy for me.

Reply:
🤣 it's not easy "period"

Reply:
After all that, I realize that my suggested name `BareCall` is confusing. What is it supposed to suggest? But I'm struggling to find a better name. Maybe `OnCompleteDispatch` or `Dispatch` or `RouterDispatch` or `Delegate` or `OnCompleteAction`. None of these are ideal, but all seem better than `BareCall`.

Reply:
or maybe just `Action` or `NonABIAction` ... argh ...

Reply:
I will up something like `OnCompleteDispatch` or `OnCompleteDelegation`

Reply:
`NonABIAction` is fine for me too

Reply:
`OnCompleteAction` ?

Reply:
deal

Reply:
> Does it even make sense to have a DEFAULT_BARE_CALLS as above, or that is too restrictive, or even dangerous?

I'm still trying to catch up, so take my feedback with a grain of salt.  Replying now because I think _some_ feedback sooner is better than waiting longer.  My initial feeling is:
* Always require caller to specify `bare_calls`.  I think the user ought to explicitly control the behavior.
* Offering factory methods and/or common combinations sounds plausible.  I'll need to consider the APIs more to provide more feedback.  Will circle back.

Reply:
I wrote up a (somewhat more updated) design of API for router in https://gist.github.com/ahangsu/78d2ed1b4edf4cd2a5e10ef6d921aef9, which might resonate with your proposal. 

My write-up includes some factor-alike method, and ensures caller to specify `bare_calls` or `OnCompleteActions` (both terminology means the same thing).